src/core/navigation/navigation.js

```diff
@@ -17,157 +15,230 @@ export default class Navigation {
-			row: index,
+			row: Math.max(this.firstRow, Math.min(this.lastRow, index)),
 			offset: offset
 		};
 	}
 
+	get columns() {
+		const columns = this.table.data.columns();
+		const indicies = [];
+		for (let i = 0, length = columns.length; i < length; i++) {
+			const column = columns[i];
+			if (column.canFocus) {
+				indicies.push(i);
+			}
+		}
+		return indicies;
+	}
+
+	get rows() {
+		return this.table.data.rows().map((r, i) => i);
+	}
+
+	get currentColumn() {
+		const columns = this.columns;
+		if (!columns.length) {
+			return -1;
+		}
+
+		const column = this.model.navigation().column;
+		const index = columns.indexOf(column);
+		return columns[Math.max(0, index)];
+	}
+
+	get nextColumn() {
+		const columns = this.columns;
+		if (!columns.length) {
+			return -1;
+		}
+
+		const column = this.model.navigation().column;
+		const index = columns.indexOf(column);
+		return index < columns.length - 1 ? columns[index + 1] : -1;
+	}
+
+	get prevColumn() {
+		const columns = this.columns;
+		if (!columns.length) {
+			return -1;
+		}
+
+		const column = this.model.navigation().column;
+		const index = columns.indexOf(column);
+		return index > 0 ? columns[index - 1] : -1;
+	}
+
+	get lastColumn() {
+		const columns = this.columns;
+		if (!columns.length) {
+			return -1;
+		}
+
+		return columns[columns.length - 1];
+	}
+
+	get firstColumn() {
+		const columns = this.columns;
+		if (!columns.length) {
+			return -1;
+		}
+
+		return columns[0];
+	}
+
+	get currentRow() {
+		const rows = this.rows;
+		if (!rows.length) {
+			return -1;
+		}
+
+		const row = this.model.navigation().row;
+		return rows[Math.max(0, row)];
+	}
+
+	get nextRow() {
+		const rows = this.rows;
+		if (!rows.length) {
+			return -1;
+		}
+
+		const row = this.model.navigation().row;
+		return row < rows.length - 1 ? rows[row + 1] : -1;
+	}
+
+	get prevRow() {
+		const rows = this.rows;
+		if (!rows.length) {
+			return -1;
+		}
+
+		const row = this.model.navigation().row;
+		return row > 0 ? rows[row - 1] : -1;
+	}
+
+	get firstRow() {
+		const rows = this.rows;
+		if (!rows.length) {
+			return -1;
+		}
+
+		return rows[0];
+	}
+
+	get lastRow() {
+		const rows = this.rows;
+		if (!rows.length) {
+			return -1;
+		}
+
+		return rows[rows.length - 1];
+	}
+
 	get commands() {
 		const model = this.model;
 		const markup = this.markup;
+		const nav = model.navigation;
+		const canExecute = () => model.edit().editMode === 'view';
+
 		const commands = {
 			goDown: new Command({
 				shortcut: 'down',
-				canExecute: () => (model.navigation().row < model.view().rows.length - 1) && model.edit().editMode == 'view',
-				execute: () => {
-					const navigationState = model.navigation();
-					if (navigationState.row == -1 && navigationState.column == -1) {
-						model.navigation({
-							column: 0,
-							row: 0
-						});
-					} else {
-						model.navigation({row: navigationState.row + 1});
-					}
-				}
+				canExecute: () => canExecute() && this.nextRow >= 0,
```

Comment:
this is common canExecute logic for all navigation commands, What name do you prefer?